ydb/core/kafka_proxy/actors/kafka_balancer_actor.cpp

```diff
@@ -0,0 +1,774 @@
+#include "kafka_balancer_actor.h"
+#include "kqp_balance_transaction.h"
+#include "ydb/core/kqp/common/simple/services.h"
+
+namespace NKafka {
+
+using namespace NKikimr;
+using namespace NKikimr::NGRpcProxy::V1;
+
+static EKafkaErrors KqpStatusToKafkaError(Ydb::StatusIds::StatusCode status) {
+    // savnik finish it
+    if (status == Ydb::StatusIds::SUCCESS) {
+        return EKafkaErrors::NONE_ERROR;
+    }
+    return EKafkaErrors::UNKNOWN_SERVER_ERROR;
+}
+
+static constexpr ui8 WAKE_UP_DELAY_SECONDS = 5;
+static constexpr ui8 WAIT_WORKING_STATE_MAX_RETRY_COUNT = 5;
+
+void TKafkaBalancerActor::Bootstrap(const NActors::TActorContext& /*ctx*/) {
+    Kqp = std::make_unique<TKqpTxHelper>("Database"); // savnik get database
+    Become(&TKafkaBalancerActor::StateWork);
+}
+
+void TKafkaBalancerActor::HandleJoinGroup(TEvKafka::TEvJoinGroupRequest::TPtr ev, const TActorContext& ctx) {
+    KAFKA_LOG_D("HandleJoinGroup request");
+
+    auto joinGroupRequest = ev->Get()->Request;
+
+    RequestType   = JOIN_GROUP;
+    CurrentStep   = STEP_NONE;
+    CorellationId = ev->Get()->CorrelationId;
+
+    GroupId  = joinGroupRequest->GroupId.value();
+    MemberId = SelfId().ToString();
+
+    Kqp->SendCreateSessionRequest(ctx);
+}
+
+void TKafkaBalancerActor::HandleSyncGroup(TEvKafka::TEvSyncGroupRequest::TPtr ev, const TActorContext& ctx) {
+    KAFKA_LOG_D("HandleSyncGroup request");
+
+    auto syncGroupRequest = ev->Get()->Request;
+
+    RequestType   = SYNC_GROUP;
+    CurrentStep   = STEP_NONE;
+    CorellationId = ev->Get()->CorrelationId;
+
+    GroupId  = syncGroupRequest->GroupId.value();
+    MemberId = syncGroupRequest->MemberId.value();
+
+    Kqp->SendCreateSessionRequest(ctx);
+}
+
+void TKafkaBalancerActor::HandleLeaveGroup(TEvKafka::TEvLeaveGroupRequest::TPtr ev, const TActorContext& ctx) {
+    KAFKA_LOG_D("HandleLeaveGroup request");
+
+    auto leaveGroupRequest = ev->Get()->Request;
+
+    RequestType   = LEAVE_GROUP;
+    CurrentStep   = STEP_NONE;
+    CorellationId = ev->Get()->CorrelationId;
+
+    GroupId  = leaveGroupRequest->GroupId.value();
+    MemberId = leaveGroupRequest->MemberId.value_or(SelfId().ToString());
+
+    Kqp->SendCreateSessionRequest(ctx);
+}
+
+void TKafkaBalancerActor::HandleHeartbeat(TEvKafka::TEvHeartbeatRequest::TPtr ev, const TActorContext& ctx) {
+    KAFKA_LOG_D("HandleHeartbeat request");
+
+    auto heartbeatRequest = ev->Get()->Request;
+
+    RequestType   = HEARTBEAT;
+    CurrentStep   = STEP_NONE;
+    CorellationId = ev->Get()->CorrelationId;
+
+    GroupId  = heartbeatRequest->GroupId.value();
+    MemberId = heartbeatRequest->MemberId.value();
+
+    Kqp->SendCreateSessionRequest(ctx);
+}
+
+
+void TKafkaBalancerActor::Handle(NKqp::TEvKqp::TEvCreateSessionResponse::TPtr& ev, const TActorContext& ctx) {
+    const TString createSessionError = "Failed to create KQP session";
+    if (!Kqp->HandleCreateSessionResponse(ev, ctx)) {
```

Comment:
if Error is TLI may be you should restart action from begin? However response with error is possible solution. Make an issue for this.